### PR TITLE
feat: Challenge 39 based on filename as encryption key

### PR DESCRIPTION
**Pull request by adarsh-a-tw**
### What kind of changes does this PR include?

-   [ ] Fixes or refactors
-   [x] A new challenge
-   [ ] Additional documentation
-   [ ] Something else

#### Description

Adds challenge 39, which is based on weak encryption due to the choice of filename as encryption key.

### Relations

Relates #975 

### Checklist:

-   [x] All the contributions made are solely the work of me and my co-authors
-   [x] I tested the changes in this PR (if applicable)
-   [x] I added unit tests to ensure my change works (when change in Java or on front-end code)
-   [x] The PR passes pre-commit hooks and automated tests

### CODE REVIEW

File: src/test/java/org/owasp/wrongsecrets/challenges/docker/Challenge39Test.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    when(resource.getContentAsString(Charset.defaultCharset()))
+        .thenReturn("YHh0aLDtW6WjFwlOgDrcRELRBIaAW9IUqR6sPy8NyYQ=");
+    when(resource.getFilename()).thenReturn("secrchallenge.md");
+  }
+
+  @Test
+  void spoilerShouldGiveAnswer() {
+    var challenge = new Challenge39(scoreCard, resource);
+    Assertions.assertThat(challenge.spoiler().solution()).isNotEmpty();
+    Assertions.assertThat(challenge.answerCorrect(challenge.spoiler().solution())).isTrue();
```
Comment (commjoen) [inline src/test/java/org/owasp/wrongsecrets/challenges/docker/Challenge39Test.java:33]: can we make sure that the decryption works? please add
```suggestion
    Assertions.assertThat(challenge.answerCorrect(challenge.spoiler().solution())).isTrue();
    Assertions.assertThat(challenge.answerCorrect("error_decryption")).isFalse();
```

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge39.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  public boolean isLimitedWhenOnlineHosted() {
+    return false;
+  }
+
+  @Override
+  public List<RuntimeEnvironment.Environment> supportedRuntimeEnvironments() {
+    return List.of(RuntimeEnvironment.Environment.DOCKER);
+  }
+
+  private String getSolution() {
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge39.java:74]: Please annotate with
```suggestion
   @SuppressFBWarnings(
      value = "CIPHER_INTEGRITY",
      justification = "This is to allow for easy ECB online decryptors")
@SuppressFBWarnings(
      value = "ECB_MODE",
      justification = "This is to allow for easy ECB online decryptors")
  private String getSolution() {
```

File: src/main/resources/explanations/challenge39_hint.adoc [hunk 3]
```diff
@@ -0,0 +1,6 @@
+This challenge can be solved by decrypting the base64 encoded secret in `secrchallenge.md`.
+
+1. Copy the contents of the `secrchallenge.md` file.
+2. Decode the copied string from base64 to UTF-8.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge39_hint.adoc:4]: can you give examples how to do that in your terminal and/or online pages?

File: src/main/resources/explanations/challenge39_hint.adoc [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+3. Decrypt it using any AES decryption tool with `secrchallenge.md` as decryption key.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge39_hint.adoc:5]: Can you give an example how to do this with terminal or online?

File: src/main/resources/explanations/challenge39_reason.adoc [hunk 5]
```diff
@@ -0,0 +1,3 @@
+*Why we should not use filename as encryption key?*
+
+There have been multiple scenarios where the contents of files were encrypted using the filename itself, which allows attackers to easily decrypt its contents.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge39_reason.adoc:3]:
```suggestion
There have been multiple scenarios where the contents of files were encrypted using the filename itself, which allows attackers to easily decrypt its contents. It is often perceived as an easy way to keep the key close to the content, without reusing the same encryption key everywhere.

But as you can tell by now: this is not a good idea. An attacker has the key the moment the file is in his posession.
```

File: src/main/resources/explanations/challenge39.adoc [hunk 6]
```diff
@@ -0,0 +1,5 @@
+=== Insecure Encryption Key - Part 1
+
+A developer encrypted a secret using https://en.wikipedia.org/wiki/Advanced_Encryption_Standard[AES] and stored it's base64 encoded value in a file. The encryption key used was same as the name of the file containing the secret.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge39.adoc:3]:
```suggestion
A develope encrypted a secret using https://en.wikipedia.org/wiki/Advanced_Encryption_Standard[AES] and stored it's base64 encoded value in a file. But where to leave the key? What about just using the filename as the encryption key instead? That way, every secret can have its own key easily! Can you find the secret?
```

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge39.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    try {
+      String encryptedText = resource.getContentAsString(Charset.defaultCharset());
+      byte[] decodedEncryptedText = Base64.getDecoder().decode(encryptedText.trim());
+      byte[] decodedKey =
+          Objects.requireNonNull(resource.getFilename()).getBytes(StandardCharsets.UTF_8);
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge39.java:79]: https://github.com/OWASP/wrongsecrets/actions/runs/6427851204/job/17454371273?pr=1023 does not like the nullpointer derefernece introduced here. maybe start with getting the key and doing null checks on the resource?

**Review by commjoen — CHANGES_REQUESTED** [review]
Hi @adarsh-a-tw you are fast! Thank you for the first challenge already!
Can you please have a look at the suggestions?
@bendehaan can you please do a textual review as well?

File: src/main/resources/explanations/challenge39.adoc [hunk 8]
```diff
@@ -0,0 +1,5 @@
+=== Insecure Encryption Key - Part 1
+
+A developer encrypted a secret using https://en.wikipedia.org/wiki/Advanced_Encryption_Standard[AES] and stored it's base64 encoded value in a file. But where to leave the key? What about just using the filename as the encryption key instead? That way, every secret can have its own key easily! Can you find the secret?
+
+The challenge file is called https://github.com/OWASP/wrongsecrets/tree/master/src/main/resources/executables/secrchallenge.md[secrchallenge.md] and can be found in the https://github.com/OWASP/wrongsecrets/tree/master/src/main/resources/executables[executables folder].
```
Comment (bendehaan) [inline src/main/resources/explanations/challenge39.adoc:5]:
```suggestion
=== Insecure Encryption Key - Part 1

A developer encrypted a secret using https://en.wikipedia.org/wiki/Advanced_Encryption_Standard[AES] and stored its base64 encoded value in a file. But where to leave the key? What about just using the filename as the encryption key instead? That way, every secret can have its own key easily! Can you find the secret?

The challenge file is called https://github.com/OWASP/wrongsecrets/tree/master/src/main/resources/executables/secrchallenge.md[secrchallenge.md] and can be found in the https://github.com/OWASP/wrongsecrets/tree/master/src/main/resources/executables[executables folder].
```

File: src/main/resources/explanations/challenge39_hint.adoc [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+1. Using an online aes decryption tool like https://www.devglan.com/online-tools/aes-encryption-decryption[https://www.devglan.com/online-tools/aes-encryption-decryption]
+- Copy the contents of the `secrchallenge.md` file and paste it in the textbox of the decryptor.
+- Make sure that the input format is `Base64` and cipher mode is `ECB`.
+- Use `secrchallenge.md` as decryption key and click on `Decrypt` to get the secret.
+
+2. Using the terminal
+- Launch the terminal while you are in the `executables` directory.
+- Type in `echo -n "secrchallenge.md" | xxd -p` to convert the plaintext key to hexadecimal key.
+- Then use the obtained decryption key to decrypt the file by typing `openssl enc -a -d -aes-128-ecb -in secrchallenge.md -K 736563726368616c6c656e67652e6d64 -out decrypted.md`
+- Copy the secret from the `decrypted.md` file in the `executables` folder.
```
Comment (bendehaan) [inline src/main/resources/explanations/challenge39_hint.adoc:12]:
```suggestion
This challenge can be solved by decrypting the base64 encoded secret in `secrchallenge.md`. You can do this either by:

1. Using an online aes decryption tool like https://www.devglan.com/online-tools/aes-encryption-decryption[https://www.devglan.com/online-tools/aes-encryption-decryption]
- Copy the contents of the `secrchallenge.md` file and paste it into the textbox of the decryptor.
- Ensure the input format is `Base64` and the cipher mode is `ECB`.
- Use `secrchallenge.md` as decryption key and click `Decrypt` to get the secret.

2. Using the terminal
- Launch the terminal while you are in the `executables` directory.
- Type in `echo -n "secrchallenge.md" | xxd -p` to convert the plaintext key to a hexadecimal key.
- Then, use the obtained decryption key to decrypt the file by typing `openssl enc -a -d -aes-128-ecb -in secrchallenge.md -K 736563726368616c6c656e67652e6d64 -out decrypted.md`
- Copy the secret from the `decrypted.md` file in the `executables` folder.
```

File: src/main/resources/explanations/challenge39_reason.adoc [hunk 10]
```diff
@@ -0,0 +1,5 @@
+*Why we should not use filename as encryption key?*
+
+There have been multiple scenarios where the contents of files were encrypted using the filename itself, which allows attackers to easily decrypt its contents. It is often perceived as an easy way to keep the key close to the content, without reusing the same encryption key everywhere.
+
+But as you can tell by now: this is not a good idea. An attacker has the key the moment the file is in his possession.
```
Comment (bendehaan) [inline src/main/resources/explanations/challenge39_reason.adoc:5]:
```suggestion
*Why should we not use the filename as the encryption key?*

There have been multiple scenarios where the contents of files were encrypted using the filename itself, which allows attackers to decrypt its contents easily. It is often perceived as an easy way to keep the key close to the content without reusing the same encryption key everywhere.

But as you can tell by now, this is not a good idea. An attacker has the key the moment the file is in his possession.
```

**Review by bendehaan — APPROVED** [review]

**Comment by bendehaan** [comment]
Looks good. Thanks for the contribution!

**Pull request merged**